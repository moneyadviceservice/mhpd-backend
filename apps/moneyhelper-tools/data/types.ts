import { useTranslation } from '@maps-digital/shared/hooks';
import { ReactNode } from 'react';

export type TranslationGroup = {
  readonly en: ReactNode;
  readonly cy: ReactNode;
};

export type TranslationGroupString = {
  readonly en: string;
  readonly cy: string;
};

export type FieldType =
  | 'select'
  | 'input-currency'
  | 'input-currency-with-select'
  | 'radio';

export interface FormFieldValidation {
  required?: boolean;
  requiredPageMessage?: string;
  requiredInputMessage?: string;
  acdlMessage?: string;
  type?: FieldType;
  rules?: ValidationRules[];
  bounds?: { lower: string; upper: string };
  userNotSeen?: boolean;
}

export interface MACResultsValidation {
  bounds: {
    lower: number;
    upper: number;
  };
}

export interface ValidationRules {
  condition: '>' | '<' | '>=' | '<=';
  otherField?: string;
  multiplier?: number;
  otherFieldMultiplier?: number;
  condition2?: '+';
  otherField2?: string;
  value?: string | number | boolean;
  rulePageMessage?: string;
  ruleInputMessage?: string;
  acdlMessage?: string;
}

export interface FormValidationObj {
  [key: string]: FormFieldValidation | undefined;
}

export interface FormFieldInputErrors {
  [key: string]: string | string[];
}

type FormFieldGroup = {
  label: string;
  text?: string;
  key: string;
  type?: GroupType;
};

export type FieldCondition = {
  field: string;
  value: string | number | boolean;
  rule: '=' | '>' | '<' | '>=' | '<=';
};

type FormFieldBase = {
  key: string;
  connectField?: string; // to be used like a foreign key
  fieldCondition?: FieldCondition; // Used on radio buttons with array of field keys to duplicate
  label: string;
  acdlLabel?: string;
  description?: string;
  type: FieldType;
  options?: {
    text: string;
    value: string;
  }[];
  expandableContent?: {
    title: string;
    text: string | ReactNode;
  };
  validation?: FormFieldValidation;
  group?: FormFieldGroup;
  addon?: string;
  topMargin?: boolean;
};

type FormFieldWithDefaultValues =
  | {
      type: 'select';
      placeholder?: string;
      placeholderInput?: never;
      placeholderSelect?: never;
      defaultInputValue?: never;
      defaultSelectValue: string | number;
      defaultRadioValue?: never;
    }
  | {
      type: 'input-currency';
      placeholder?: string;
      placeholderInput?: never;
      placeholderSelect?: never;
      defaultInputValue?: string | number;
      defaultSelectValue?: never;
      defaultRadioValue?: never;
    }
  | {
      type: 'input-currency-with-select';
      placeholder?: never;
      placeholderInput?: string;
      placeholderSelect?: string;
      defaultInputValue?: string | number;
      defaultSelectValue: string | number;
      defaultRadioValue?: never;
    }
  | {
      type: 'radio';
      placeholder?: string;
      placeholderInput?: never;
      placeholderSelect?: never;
      defaultInputValue?: never;
      defaultSelectValue?: never;
      defaultRadioValue?: string | number;
    };

export type FormField = FormFieldBase & FormFieldWithDefaultValues;

export type TabResultContent = {
  success: {
    title: string;
    content: string;
  };
  error: {
    title: string;
    content: string;
  };
  subHeading: string;
};

export type SummaryCalc = 'sub' | 'add' | 'result' | 'exclude';

export type Summary = {
  label: string;
  unit: 'pounds' | 'months';
  calc?: SummaryCalc;
};

type FormContent = {
  content: string | ReactNode;
  fields: FormField[];
  result?: never;
};

type ResultContent = {
  content?: never;
  fields?: never;
  result: TabResultContent;
};

export type StepData = {
  key: string;
  heading: string;
  buttonText: string;
} & (FormContent | ResultContent);

export type TabData = {
  key?: string;
  linkText?: string;
  summary?: Summary;
  heading: string;
} & (FormContent | ResultContent);

export type TabPageData = {
  toolHeading: string;
  summaryHeading: string;
  tabs: TabData[];
};

export type StepPageData = {
  toolHeading: string;
  steps: StepData[];
};

export type FormData = {
  [key: string]: string | string[] | undefined;
};

export type DefaultValues = {
  [key: string]: string | number;
};

export type ConditionalFields = Set<string>;

export type GroupedFieldItem = {
  name: string;
  value: number;
  url: string;
};

export enum GroupType {
  HEADING = 'heading',
  EXPANDABLE = 'expandable',
}

type ErrorMessageObj = {
  required?: {
    field?: string;
    page?: string;
    acdlMessage?: string;
  };
  condition?: {
    field?: string;
    page?: string;
    acdlMessage?: string;
  };
};

export type ErrorMessageFunc = (
  z: ReturnType<typeof useTranslation>['z'],
) => ErrorMessageObj;

export type Errors = Record<string, string[]> | null;
