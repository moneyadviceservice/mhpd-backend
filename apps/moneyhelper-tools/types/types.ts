import { ReactNode } from 'react';

export type Question = {
  questionNbr: number;
  group: string;
  definition?: string | ReactNode;
  title: string;
  type: string;
  subType?: string;
  target?: string;
  description?: string | ReactNode;
  answers: Answer[];
  inputProps?: {
    maxLimit?: number;
    labelValue?: string;
  };
  errors?: Record<string, string | ReactNode>;
};

export type Answer = {
  text: string;
  subtext?: ReactNode;
  score?: number;
  unselectedScore?: number;
  clearAll?: boolean;
  showLinksIfSelected?: boolean;
  links?: Links[];
  unselectedAnswerLinks?: Links[];
};

export type Links = {
  title: string;
  link: string;
  type: string;
  description?: string;
};

export type ErrorType = {
  question: number;
  message: string;
};

export type FormContentAnlyticsData = {
  pageName: string;
  pageTitle: string;
  toolName: string;
  stepNames: string[] | string;
  categoryLevels?: string[];
};

export enum DataPath {
  MidLifeMot = 'midlife-mot',
  CreditRejection = 'credit-rejection',
  CreditOptions = 'credit-options',
  PensionType = 'pension-type',
  BabyCostCalculator = 'baby-cost-calculator',
  WorkplacePensionCalculator = 'workplace-pension-calculator',
  CashInChunksCalculator = 'cash-in-chunks-calculator',
  TakeWholePot = 'take-whole-pot',
  GuaranteedIncomeEstimator = 'guaranteed-income-estimator',
  LeavePotUntouched = 'leave-pot-untouched',
  AjustableIncomeEstimator = 'ajustable-income-estimator',
  MortgageAffordability = 'mortgage-affordability-calculator',
  SavingsCalculator = 'savings-calculator',
  BabyMoneyTimeline = 'baby-money-timeline',
}

export enum UrlPath {
  MidLifeMot = 'mid-life-mot',
  CreditRejection = 'credit-rejection',
  CreditOptions = 'credit-options',
  PensionType = 'pension-type',
  BabyCostCalculator = 'baby-cost-calculator',
  WorkplacePensionCalculator = 'workplace-pension-calculator',
  CashInChunksCalculator = 'cash-in-chunks-calculator',
  TakeWholePot = 'take-whole-pot',
  GuaranteedIncomeEstimator = 'guaranteed-income-estimator',
  LeavePotUntouched = 'leave-pot-untouched',
  AjustableIncomeEstimator = 'ajustable-income-estimator',
  MortgageAffordability = 'mortgage-affordability-calculator',
  SavingsCalculator = 'savings-calculator',
  BabyMoneyTimeline = 'baby-money-timeline',
}

export type Group = {
  title: string;
  group: string;
  descritionScoreOne?: string;
  descritionScoreTwo?: string;
  descritionScoreThree?: string;
};

export type TranslationGroup = {
  readonly en: ReactNode;
  readonly cy: ReactNode;
};

export type TranslationGroupString = {
  readonly en: string;
  readonly cy: string;
};

export type Condition = {
  question: string;
  answer: string;
  arithmeticOperator?: string;
};
