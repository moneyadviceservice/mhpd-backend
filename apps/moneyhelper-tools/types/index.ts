export type {
  Question,
  Answer,
  Links,
  ErrorType,
  FormContentAnlyticsData,
  Group,
  TranslationGroup,
  TranslationGroupString,
  Condition,
} from './types';

export { DataPath, UrlPath } from './types';
