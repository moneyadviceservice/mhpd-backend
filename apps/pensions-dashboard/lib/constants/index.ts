import Cookies from 'cookies';

export const SITE_TITLE = 'MoneyHelper Pensions Dashboard';

export const NUMBER_WORDS = [
  { en: 'zero', cy: 'sero' },
  { en: 'one', cy: 'un' },
  { en: 'two', cy: 'dau' },
  { en: 'three', cy: 'tri' },
  { en: 'four', cy: 'pedwar' },
  { en: 'five', cy: 'pump' },
  { en: 'six', cy: 'chwech' },
  { en: 'seven', cy: 'saith' },
  { en: 'eight', cy: 'wyth' },
  { en: 'nine', cy: 'naw' },
  { en: 'ten', cy: 'deg' },
  { en: 'eleven', cy: 'un ar ddeg' },
  { en: 'twelve', cy: 'deuddeg' },
];

export const PROTOCOL =
  process?.env.NODE_ENV === 'development' ? 'http://' : 'https://';

export const COOKIE_OPTIONS = {
  path: '/',
  httpOnly: true,
  sameSite: 'lax',
} as Cookies.SetOption;

export enum MatchType {
  DEFN = 'DEFN',
  POSS = 'POSS',
  CONT = 'CONT',
  NEW = 'NEW',
  SYS = 'SYS',
}

export enum PensionType {
  AVC = 'AVC',
  CB = 'CB',
  CDC = 'CDC',
  DB = 'DB',
  DC = 'DC',
  HYB = 'HYB',
  SP = 'SP',
}

export enum PensionOrigin {
  A = 'A',
  PC = 'PC',
  PM = 'PM',
  PT = 'PT',
  WC = 'WC',
  WM = 'WM',
  WT = 'WT',
}

export enum PensionStatus {
  A = 'A',
  I = 'I',
  IPPF = 'IPPF',
  IWU = 'IWU',
}

export enum AlternateNameType {
  FOR = 'FOR',
  OTH = 'OTH',
}

export enum EmployerStatus {
  C = 'C',
  H = 'H',
}

export enum InformationType {
  C_AND_C = 'C_AND_C',
  SIP = 'SIP',
  IMP = 'IMP',
  ANR = 'ANR',
  SP = 'SP',
}

export enum RetrievalStatus {
  NEW = 'NEW',
  RETRIEVAL_REQUESTED = 'RETRIEVAL_REQUESTED',
  RETRIEVAL_COMPLETE = 'RETRIEVAL_COMPLETE',
  RETRIEVAL_FAILED = 'RETRIEVAL_FAILED',
}

export enum UsageType {
  A = 'A',
  M = 'M',
  N = 'N',
  S = 'S',
  W = 'W',
}

export enum IllustrationType {
  ERI = 'ERI',
  AP = 'AP',
  DB = 'DB',
}

export enum BenefitType {
  AVC = 'AVC',
  CBL = 'CBL',
  CBS = 'CBS',
  CDI = 'CDI',
  CDL = 'CDL',
  DB = 'DB',
  DBL = 'DBL',
  DC = 'DC',
  SP = 'SP',
}

export enum UnavailableReason {
  ANO = 'ANO',
  DB = 'DB',
  DBC = 'DBC',
  DCC = 'DCC',
  DCHA = 'DCHA',
  DCHP = 'DCHP',
  DCSM = 'DCSM',
  MEM = 'MEM',
  NEW = 'NEW',
  NET = 'NET',
  PPF = 'PPF',
  TRN = 'TRN',
  WU = 'WU',
}

export enum IllustratonWarning {
  AVC = 'AVC',
  CUR = 'CUR',
  DEF = 'DEF',
  FAS = 'FAS',
  PEO = 'PEO',
  PNR = 'PNR',
  PSO = 'PSO',
  SPA = 'SPA',
  TVI = 'TVI',
  INP = 'INP',
}

export enum CalculationMethod {
  BS = 'BS',
  CBI = 'CBI',
  SMPI = 'SMPI',
}

export enum LumpSumAmountType {
  CSH = 'CSH',
  CSHL = 'CSHL',
  CSHN = 'CSHN',
}

export enum RecurringAmountType {
  INC = 'INC',
  INCL = 'INCL',
  INCN = 'INCN',
}

export enum AmountNotProvidedReason {
  SML = 'SML',
}

export enum ContactMethods {
  TELEPHONE = 'number',
  EMAIL = 'email',
  POSTAL_NAME = 'postalName',
  WEB_CONTACTS = 'url',
}
