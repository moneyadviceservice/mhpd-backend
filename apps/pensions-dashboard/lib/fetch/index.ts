export * from './get-all-pensions';
export * from './get-pension-detail';
export * from './get-pensions-data';
export * from './get-pensions-overview';
export * from './post-pensions-data';
export * from './post-redirect-details';
