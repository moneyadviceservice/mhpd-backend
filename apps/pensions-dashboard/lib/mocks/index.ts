export const mockPensionsData = {
  pensionPolicies: [
    {
      pensionArrangements: [
        {
          externalPensionPolicyId: 'D2007659822',
          externalAssetId: 'b3cd4ebd-8a49-4ebc-a15e-1e79dff1e5ad',
          matchType: 'DEFN',
          schemeName: 'Master Trust Workplace 0887',
          contactReference: 'Ref/rb-dr-bd-sb-08',
          pensionType: 'DC',
          pensionOrigin: 'WM',
          pensionStatus: 'I',
          startDate: '2011-05-16',
          retirementDate: '2025-08-23',
          dateOfBirth: '1979-09-18',
          pensionAdministrator: {
            name: 'Pension Admin Highland',
            contactMethods: [
              {
                preferred: true,
                contactMethodDetails: {
                  url: 'https://www.highlandadmin.co.uk',
                },
              },
              {
                preferred: true,
                contactMethodDetails: {
                  email: 'mastertrust@highlandadmin.com',
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 800873434',
                  usage: ['M'],
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  postalName: 'Pension Admin Highland',
                  line1: '1 Travis Avenue',
                  line2: 'Main Street',
                  line3: 'Liverpool',
                  postcode: 'M16 0QG',
                  countryCode: 'GB',
                },
              },
            ],
          },
          employmentMembershipPeriods: [
            {
              employerName: 'Borough Finance Centre',
              membershipStartDate: '2011-05-16',
            },
          ],
          benefitIllustrations: [
            {
              illustrationComponents: [
                {
                  illustrationType: 'ERI',
                  unavailableReason: 'DCC',
                  benefitType: 'DC',
                  calculationMethod: 'SMPI',
                  survivorBenefit: true,
                  dcPot: 540500,
                  safeguardedBenefit: true,
                },
                {
                  illustrationType: 'AP',
                  unavailableReason: 'DCC',
                  benefitType: 'DC',
                  amountType: 'INC',
                  calculationMethod: 'SMPI',
                  dcPot: 540500,
                  survivorBenefit: true,
                  safeguardedBenefit: true,
                },
              ],
              illustrationDate: '2024-01-01',
            },
          ],
        },
        {
          externalPensionPolicyId: 'D9267759822',
          externalAssetId: 'e4f88759-5c0b-4148-82aa-9c6611914e9c',
          matchType: 'DEFN',
          schemeName: 'Workers Trust Local 701',
          contactReference: 'Ref/rb-dr-bd-sb-08',
          pensionType: 'DC',
          pensionOrigin: 'WM',
          pensionStatus: 'I',
          startDate: '2002-07-03',
          retirementDate: '2025-08-23',
          dateOfBirth: '1979-09-18',
          pensionAdministrator: {
            name: 'Pension For Everyone',
            contactMethods: [
              {
                preferred: false,
                contactMethodDetails: {
                  url: 'https://www.everypension.co.uk',
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  email: 'query@everypension.com',
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 800093434',
                  usage: ['M'],
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  postalName: 'Pension For Everyone',
                  line1: '12 Mike Close',
                  line2: 'Main Street',
                  line3: 'Newcastle',
                  postcode: 'S16 3BG',
                  countryCode: 'GB',
                },
              },
            ],
          },
          benefitIllustrations: [
            {
              illustrationComponents: [
                {
                  illustrationType: 'ERI',
                  unavailableReason: 'DCHA',
                  benefitType: 'DC',
                  calculationMethod: 'SMPI',
                  survivorBenefit: false,
                  dcPot: 311011,
                  safeguardedBenefit: false,
                },
                {
                  illustrationType: 'AP',
                  unavailableReason: 'DCHA',
                  benefitType: 'DC',
                  amountType: 'INC',
                  calculationMethod: 'SMPI',
                  dcPot: 311011,
                  survivorBenefit: false,
                  safeguardedBenefit: false,
                },
              ],
              illustrationDate: '2024-01-09',
            },
          ],
        },
        {
          matchType: 'DEFN',
          schemeName: 'State Pension',
          pensionType: 'SP',
          retirementDate: '2042-02-23',
          externalAssetId: '7f0763a9-ac18-43c3-b2e7-723a74eba292',
          benefitIllustrations: [
            {
              illustrationDate: '2024-08-24',
              illustrationComponents: [
                {
                  benefitType: 'SP',
                  payableDetails: {
                    amountType: 'INC',
                    payableDate: '2042-02-23',
                    annualAmount: 11502,
                    monthlyAmount: 958.5,
                    increasing: true,
                  },
                  illustrationType: 'ERI',
                  calculationMethod: 'BS',
                },
              ],
            },
            {
              illustrationDate: '2024-08-24',
              illustrationComponents: [
                {
                  benefitType: 'SP',
                  payableDetails: {
                    amountType: 'INC',
                    payableDate: '2042-02-23',
                    annualAmount: 11502,
                    monthlyAmount: 958.5,
                    increasing: true,
                  },
                  illustrationType: 'AP',
                  calculationMethod: 'BS',
                },
              ],
            },
          ],
          pensionAdministrator: {
            name: 'DWP',
            contactMethods: [
              {
                preferred: false,
                contactMethodDetails: {
                  postalName: 'Freepost DWP',
                  line1: 'Pensions Service 3',
                  countryCode: 'GB',
                },
              },
              {
                preferred: true,
                contactMethodDetails: {
                  url: 'https://www.gov.uk/future-pension-centre',
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 8007310175',
                  usage: ['M', 'W'],
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 8007310176',
                  usage: ['M'],
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 8007310456',
                  usage: ['W'],
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 1912182051',
                  usage: ['N'],
                },
              },
              {
                preferred: false,
                contactMethodDetails: {
                  number: '+44 1912183600',
                  usage: ['N'],
                },
              },
            ],
          },
          additionalDataSources: [
            {
              url: 'https://www.gov.uk/check-state-pension',
              informationType: 'SP',
            },
          ],
          statePensionMessageEng: 'State pension message in English.',
          statePensionMessageWelsh: 'Neges pensiwn gwladol yn Saesneg.',
        },
      ],
    },
  ],
  peiInformation: {
    peiRetrievalComplete: false,
    peiData: [
      {
        pei: '4572fe02-ed14-4738-b6c1-7b71cfd09c16:e4f88759-5c0b-4148-82aa-9c6611914e9c',
        description: 'Workers Trust Local 701',
        retrievalStatus: 'RETRIEVAL_COMPLETE',
      },
      {
        pei: '7075aa11-10ad-4b2f-a9f5-1068e79119bf:b3cd4ebd-8a49-4ebc-a15e-1e79dff1e5ad',
        description: 'Master Trust Workplace 0887',
        retrievalStatus: 'RETRIEVAL_COMPLETE',
      },
      {
        pei: '7075aa11-10ad-4b2f-a9f5-1068e79119bf:7f0763a9-ac18-43c3-b2e7-723a74eba292',
        description: 'State Pension Details',
        retrievalStatus: 'RETRIEVAL_COMPLETE',
      },
    ],
  },
  pensionsDataRetrievalComplete: true,
};

export const mockPensionDetail = [
  mockPensionsData.pensionPolicies[0].pensionArrangements[0],
  mockPensionsData.pensionPolicies[0].pensionArrangements[1],
];

export const mockIncompletePensions = [
  mockPensionsData.pensionPolicies[0].pensionArrangements[0],
  mockPensionsData.pensionPolicies[0].pensionArrangements[1],
];

export const mockConfirmedPensions = [
  mockPensionsData.pensionPolicies[0].pensionArrangements[2],
];

export const mockPolicyPension = [
  {
    ...mockPensionsData.pensionPolicies[0].pensionArrangements[0],
    benefitIllustrations: undefined,
  },
];

export const mockUnconfirmedPensions = [
  {
    externalPensionPolicyId: 'D2007659822',
    externalAssetId: 'b3cd4ebd-8a49-4ebc-a15e-1e79dff1e5ad',
    matchType: 'POSS',
    schemeName: 'Master Trust Workplace 0887',
    contactReference: 'Ref/rb-dr-bd-sb-08',
    pensionType: 'DC',
    pensionOrigin: 'WM',
    pensionStatus: 'I',
    startDate: '2011-05-16',
    retirementDate: '2025-08-23',
    dateOfBirth: '1979-09-18',
    pensionAdministrator: {
      name: 'Pension Admin Highland',
      contactMethods: [
        {
          preferred: true,
          contactMethodDetails: { url: 'https://www.highlandadmin.co.uk' },
        },
        {
          preferred: true,
          contactMethodDetails: { email: 'mastertrust@highlandadmin.com' },
        },
        {
          preferred: false,
          contactMethodDetails: { number: '+44 800873434', usage: ['M'] },
        },
        {
          preferred: false,
          contactMethodDetails: {
            postalName: 'Pension Admin Highland',
            line1: '1 Travis Avenue',
            line2: 'Main Street',
            line3: 'Liverpool',
            postcode: 'M16 0QG',
            countryCode: 'GB',
          },
        },
      ],
    },
    employmentMembershipPeriods: [
      {
        employerName: 'Borough Finance Centre',
        membershipStartDate: '2011-05-16',
      },
    ],
    benefitIllustrations: [
      {
        illustrationComponents: [
          {
            illustrationType: 'ERI',
            unavailableReason: 'DCC',
            benefitType: 'DC',
            calculationMethod: 'SMPI',
            survivorBenefit: true,
            dcPot: 540500,
            safeguardedBenefit: true,
          },
          {
            illustrationType: 'AP',
            unavailableReason: 'DCC',
            benefitType: 'DC',
            amountType: 'INC',
            calculationMethod: 'SMPI',
            dcPot: 540500,
            survivorBenefit: true,
            safeguardedBenefit: true,
          },
        ],
        illustrationDate: '2024-01-01',
      },
    ],
  },
];
