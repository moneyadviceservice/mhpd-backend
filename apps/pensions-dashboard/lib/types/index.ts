import {
  AlternateNameType,
  AmountNotProvidedReason,
  BenefitType,
  CalculationMethod,
  EmployerStatus,
  IllustrationType,
  IllustratonWarning,
  InformationType,
  LumpSumAmountType,
  MatchType,
  PensionOrigin,
  PensionStatus,
  PensionType,
  RecurringAmountType,
  RetrievalStatus,
  UnavailableReason,
  UsageType,
} from '../constants';

export type ErrorResponse = {
  error: string;
  message: string;
};

export type PensionData = {
  pensionPolicies: PensionPolicy[];
  peiInformation: PeiInformationModel;
  pensionsDataRetrievalComplete: boolean;
};

type PensionPolicy = {
  pensionArrangements: PensionArrangement[];
};

export type PensionArrangement = {
  externalPensionPolicyId: string;
  externalAssetId: string;
  matchType: MatchType;
  schemeName: string;
  alternateSchemeNames?: AlternateScheme[];
  contactReference: string;
  pensionType: PensionType;
  pensionOrigin?: PensionOrigin;
  pensionStatus?: PensionStatus;
  startDate?: string;
  retirementDate?: string;
  dateOfBirth?: string;
  statePensionMessageEng?: string;
  statePensionMessageWelsh?: string;
  contributionsFromMultipleEmployers: boolean;
  pensionAdministrator: PensionAdministrator;
  employmentMembershipPeriods?: EmploymentMembershipPeriod[];
  benefitIllustrations?: BenefitIllustration[];
  additionalDataSources?: AdditionalDataSource[];
};

type AlternateScheme = {
  name: string;
  alternateNameType: AlternateNameType;
};

type PensionAdministrator = {
  name: string;
  contactMethods: ContactMethod[];
};

export type ContactMethod = {
  preferred: boolean;
  contactMethodDetails: EmailAddress | PhoneNumber | PostalAddress | Website;
};

type EmploymentMembershipPeriod = {
  employerName: string;
  employerStatus: EmployerStatus;
  membershipStartDate: string;
  membershipEndDate: string;
};

export type BenefitIllustration = {
  illustrationComponents: BenefitIllustrationComponent[];
  illustrationDate: string;
};

export type AdditionalDataSource = {
  url: string;
  informationType: InformationType;
};

type PeiInformationModel = {
  peiRetrievalComplete: boolean;
  peiData: PeiData[];
};

type PeiData = {
  pei: string;
  description: string;
  retrievalStatus: RetrievalStatus;
};

export type EmailAddress = {
  email: string;
};

export type PhoneNumber = {
  number: string;
  usage: UsageType[];
};

export type PostalAddress = {
  postalName: string;
  line1: string;
  line2?: string;
  line3?: string;
  line4?: string;
  line5?: string;
  postcode: string;
  countryCode: string;
};

export type Website = {
  url: string;
};

export type BenefitIllustrationComponent = {
  illustrationType: IllustrationType;
  unavailableReason?: UnavailableReason;
  benefitType: BenefitType;
  calculationMethod: CalculationMethod;
  payableDetails: PayableDetails;
  dcPot?: number;
  survivorBenefit?: boolean;
  safeguardedBenefit?: boolean;
  illustratonWarnings?: IllustratonWarning[];
};

type PayableDetails =
  | RecurringIncomeDetails
  | LumpSumDetails
  | AmountNotProvidedDetails;

export type RecurringIncomeDetails = {
  amountType: RecurringAmountType;
  annualAmount: number;
  monthlyAmount: number;
  lastPaymentDate?: string;
  payableDate: string;
  increasing: boolean;
};

type LumpSumDetails = {
  amountType: LumpSumAmountType;
  amount: number;
  payableDate: string;
};

export type AmountNotProvidedDetails = {
  reason: AmountNotProvidedReason;
  payableDate: string;
  lastPaymentDate?: string;
};

export type PensionsOverviewModel = {
  totalPensions: number;
  confirmedPensions: PensionArrangement[];
  unconfirmedPensions: PensionArrangement[];
  incompletePensions: PensionArrangement[];
};

export type PostResponseType = {
  requestID: string;
  rqp: string;
  scope: string;
  responseType: string;
  prompt: string;
  service: string;
  codeChallengeMethod: string;
  codeChallenge: string;
  clientId: string;
  code_verifier: string;
  redirectTargetUrl: string;
};
