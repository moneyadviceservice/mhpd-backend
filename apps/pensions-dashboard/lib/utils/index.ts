export * from './formatDate';
export * from './formatPhoneNumber';
export * from './getLatestIllustration';
export * from './getRetirementAge';
export * from './getUserSessionFromCookies';
export * from './getUnavailableText';
export * from './numberToWords';
export * from './pension';
export * from './sortPensions';
export * from './toCurrency';
export * from './transformPensionData';
export * from './contactMethods';
