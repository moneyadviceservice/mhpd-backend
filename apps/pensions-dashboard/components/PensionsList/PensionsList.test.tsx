import React from 'react';
import { render } from '@testing-library/react';
import { PensionsList } from '.';
import { Icon, IconType } from '@maps-react/common/components/Icon';
import { PensionArrangement } from '../../lib/types';

const pensions = [
  {
    externalAssetId: '1',
    matchType: 'DEFN',
    schemeName: 'State Pension',
    pensionType: 'SP',
    contributionsFromMultipleEmployers: false,
    pensionAdministrator: {
      name: 'DWP',
      contactMethods: [
        {
          preferred: false,
          contactMethods: [
            {
              preferred: true,
              contactMethodDetails: {
                number: '+123 1111111111',
                usage: ['A', 'M'],
              },
            },
          ],
        },
      ],
    },
  },
  {
    externalAssetId: '2',
    matchType: 'DEFN',
    schemeName: 'Your Pension DC Master Trust',
    pensionType: 'DC',
    contributionsFromMultipleEmployers: false,
    pensionAdministrator: {
      name: 'Your Pension',
      contactMethods: [
        {
          preferred: true,
          contactMethodDetails: {
            number: '+123 1111111111',
            usage: ['A', 'M'],
          },
        },
      ],
    },
  },
] as PensionArrangement[];

describe('PensionsList', () => {
  it('renders correctly', () => {
    const { container } = render(
      <PensionsList
        pensions={pensions}
        icon={<Icon type={IconType.TICK_GREEN} />}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
