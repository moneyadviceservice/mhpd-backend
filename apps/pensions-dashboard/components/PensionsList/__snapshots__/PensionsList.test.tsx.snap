// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PensionsList renders correctly 1`] = `
<div>
  <ul>
    <li
      class="flex items-center gap-2 mt-4"
    >
      <span
        class="w-[24px] h-[24px]"
      >
        <svgr-mock
          classname=""
        />
      </span>
      State Pension
    </li>
    <li
      class="flex items-center gap-2 mt-4"
    >
      <span
        class="w-[24px] h-[24px]"
      >
        <svgr-mock
          classname=""
        />
      </span>
      Your Pension
    </li>
  </ul>
</div>
`;
